File: Arrays & Strings/125. Valid Palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if(len(s) == 0 or len(s) == 1):
            return True
        # s = ''.join(filter(str.isalnum, s)).lower()  #removing the non alphanum characters
        l, r = 0, len(s) - 1
        while(l < r):
            while(not s[l].isalnum() and l < r):   #skipping the non alphanum characters 
                l += 1
            while(not s[r].isalnum() and l < r):
                r -= 1
            if(s[l].lower() != s[r].lower()):
                return False        
            l += 1
            r -= 1
        return True
        

if __name__ == "__main__":
    sol = Solution()
    s = "A man, a plan, a canal: Panama"
    print(sol.isPalindrome(s))

File: Two Pointers/125. Valid Palindrome.cpp
/* 125. Valid Palindrome
 */

#include <bits/stdc++.h>

class Solution
{
public:
    bool isPalindrome(std::string s)
    {
        if(s.size() == 0 or s.size() == 1) {
            return true;
        }

        int l = 0, r = s.size() - 1;

        while (l < r)
        {
            while (!isalnum(s[l]) && l < r)
                l++;
            while (!isalnum(s[r]) && l < r)
                r--;

            if (std::tolower(s[l]) != std::tolower(s[r]))
                return false;
            
            l++;
            r--;
        }

        return true;
    }
};

int main()
{
    std::string s = "A man, a plan, a canal: Panama";

    Solution sol;
    std::cout << sol.isPalindrome(s) << std::endl;

    return 0;
}